Panic with OffsetFetchRequest

I'm using franz-go in some unit tests to watch produced and committed offsets and I'm occasionally seeing the same panic in our CI environment in tests (it occurs maybe 10% of the time or so):

```
panic: interface conversion: kmsg.Request is *kmsg.OffsetFetchRequest, not *kgo.pinReq [recovered]
	panic: interface conversion: kmsg.Request is *kmsg.OffsetFetchRequest, not *kgo.pinReq
goroutine 34 [running]:
testing.tRunner.func1.2({0x13619e0, 0xc0000fe480})
	/usr/local/go/src/testing/testing.go:1396 +0x372
testing.tRunner.func1()
	/usr/local/go/src/testing/testing.go:1399 +0x5f0
panic({0x13619e0, 0xc0000fe480})
	/usr/local/go/src/runtime/panic.go:890 +0x262
github.com/twmb/franz-go/pkg/kgo.(*Client).handleShardedReq.func2({0x0, {0x1666848, 0xc0000a0540}, {0x0, 0x0}})
	/go/pkg/mod/github.com/twmb/franz-go@v1.11.2/pkg/kgo/client.go:1804 +0xd05
github.com/twmb/franz-go/pkg/kgo.(*Client).handleShardedReq(0xc0005d2000, {0x1660918?, 0xc000036380}, {0x1666848?, 0xc0000a0540})
	/go/pkg/mod/github.com/twmb/franz-go@v1.11.2/pkg/kgo/client.go:1859 +0xed1
github.com/twmb/franz-go/pkg/kgo.(*Client).shardedRequest(0xc0005d2000, {0x1660950, 0xc00013a008}, {0x1666848, 0xc0000a0540})
	/go/pkg/mod/github.com/twmb/franz-go@v1.11.2/pkg/kgo/client.go:994 +0x811
github.com/twmb/franz-go/pkg/kgo.(*Client).Request(0xc0000add90?, {0x1660950, 0xc00013a008}, {0x1666848, 0xc0000a0540})
	/go/pkg/mod/github.com/twmb/franz-go@v1.11.2/pkg/kgo/client.go:778 +0x71
github.com/twmb/franz-go/pkg/kmsg.(*OffsetFetchRequest).RequestWith(...)
	/go/pkg/mod/github.com/twmb/franz-go/pkg/kmsg@v1.2.0/generated.go:10894
```

The code that runs the `RequestWith` call looks like this:

```
var resp *kmsg.OffsetFetchResponse
for {
  req := kmsg.NewOffsetFetchRequest()
  req.Group = c.ConsumerGroup
  c.mu.Lock()
  for tp := range c.offsets {
    req.Topics = append(req.Topics, kmsg.OffsetFetchRequestTopic{
      Topic:      tp.Topic,
      Partitions: []int32{tp.Partition},
    })
  }
  c.mu.Unlock()

  var err error
  resp, err = req.RequestWith(context.Background(), client) // <--- THIS LINE LEADS TO THE PANIC
  var ke *kerr.Error
  if err != nil && errors.As(err, &ke) {
    if ke.Code == 15 {
      time.Sleep(100 * time.Millisecond)
      continue
    }
  }
  require.Nil(tb, err)
  break
}
```

I'm on the latest version (1.11.2) on Go 1.19.  I'm being careful to not reuse the request object between tries.  Is this a bug or just user error?

Something that might be significant is that the unit tests use a single-broker kafka deployment.

--- a/.github/workflows/lint-and-test.yml
+++ b/.github/workflows/lint-and-test.yml
@@ -27,3 +27,55 @@ jobs:
       - uses: golangci/golangci-lint-action@v3
         with:
           version: latest
+
+  integration-test-kafka:
+    if: github.repository == 'twmb/franz-go'
+    needs: golangci
+    runs-on: ubuntu-latest
+    name: "integration test kafka"
+    container: golang:1.19.2
+    services:
+      kafka:
+        image: bitnami/kafka:latest
+        ports:
+          - 9092:9092
+        env:
+          KAFKA_ENABLE_KRAFT: yes
+          KAFKA_CFG_PROCESS_ROLES: controller,broker
+          KAFKA_CFG_CONTROLLER_LISTENER_NAMES: CONTROLLER
+          KAFKA_CFG_LISTENERS: PLAINTEXT://:9092,CONTROLLER://:9093
+          KAFKA_CFG_LISTENER_SECURITY_PROTOCOL_MAP: CONTROLLER:PLAINTEXT,PLAINTEXT:PLAINTEXT
+          KAFKA_CFG_CONTROLLER_QUORUM_VOTERS: 1@127.0.0.1:9093
+          # Set this to "PLAINTEXT://127.0.0.1:9092" if you want to run this container on localhost via Docker
+          KAFKA_CFG_ADVERTISED_LISTENERS: PLAINTEXT://kafka:9092
+          KAFKA_CFG_BROKER_ID: 1
+          ALLOW_PLAINTEXT_LISTENER: yes
+          KAFKA_KRAFT_CLUSTER_ID: XkpGZQ27R3eTl3OdTm2LYA # 16 byte base64-encoded UUID
+    steps:
+      - uses: actions/checkout@v3
+      - run: go test ./...
+        env:
+          KGO_TEST_RF: 1
+          KGO_SEEDS: kafka:9092
+          KGO_TEST_UNSAFE: true
+          KGO_TEST_STABLE_FETCH: true
+
+  integration-test-redpanda:
+    if: github.repository == 'twmb/franz-go'
+    needs: golangci
+    runs-on: ubuntu-latest
+    name: "integration test redpanda"
+    container: golang:1.19.2
+    services:
+      redpanda:
+        image: vectorized/redpanda-nightly:latest
+        ports:
+          - 9092:9092
+        env:
+          REDPANDA_ADVERTISE_KAFKA_ADDRESS: redpanda:9092
+    steps:
+      - uses: actions/checkout@v3
+      - run: go test ./...
+        env:
+          KGO_TEST_RF: 1
+          KGO_SEEDS: redpanda:9092

--- a/pkg/kgo/client.go
+++ b/pkg/kgo/client.go
@@ -1657,14 +1657,6 @@ type issueShard struct {
 
 // sharder splits a request.
 type sharder interface {
-	// If a request originally was not batched, then the protocol switched
-	// to being batched, we always try batched first then fallback.
-	//
-	// Requests that make this switch should always return pinReq requests,
-	// and we must unpack the pinReq to return to end users / use
-	// internally.
-	unpackPinReq() bool
-
 	// shard splits a request and returns the requests to issue tied to the
 	// brokers to issue the requests to. This can return an error if there
 	// is some pre-loading that needs to happen. If an error is returned,
@@ -1800,8 +1792,10 @@ func (cl *Client) handleShardedReq(ctx context.Context, req kmsg.Request) ([]Res
 		for i := range issues {
 			myIssue := issues[i]
 			myUnderlyingReq := myIssue.req
-			if sharder.unpackPinReq() {
-				myUnderlyingReq = myIssue.req.(*pinReq).Request
+			var isPinned bool
+			if pinned, ok := myIssue.req.(*pinReq); ok {
+				myUnderlyingReq = pinned.Request
+				isPinned = true
 			}
 
 			if myIssue.err != nil {
@@ -1837,7 +1831,7 @@ func (cl *Client) handleShardedReq(ctx context.Context, req kmsg.Request) ([]Res
 				backoff := cl.cfg.retryBackoff(tries)
 				if err != nil &&
 					(retryTimeout == 0 || time.Now().Add(backoff).Sub(start) < retryTimeout) &&
-					(reshardable && sharder.unpackPinReq() && errors.Is(err, errBrokerTooOld) || cl.shouldRetry(tries, err) && cl.waitTries(ctx, backoff)) {
+					(reshardable && isPinned && errors.Is(err, errBrokerTooOld) || cl.shouldRetry(tries, err) && cl.waitTries(ctx, backoff)) {
 					// Non-reshardable re-requests just jump back to the
 					// top where the broker is loaded. This is the case on
 					// requests where the original request is split to
@@ -2114,8 +2108,6 @@ func (l *unknownErrShards) collect(mkreq, mergeParts interface{}) []issueShard {
 // handles sharding ListOffsetsRequest
 type listOffsetsSharder struct{ *Client }
 
-func (*listOffsetsSharder) unpackPinReq() bool { return false }
-
 func (cl *listOffsetsSharder) shard(ctx context.Context, kreq kmsg.Request, _ error) ([]issueShard, bool, error) {
 	req := kreq.(*kmsg.ListOffsetsRequest)
 
@@ -2241,8 +2233,6 @@ func (*listOffsetsSharder) merge(sresps []ResponseShard) (kmsg.Response, error) 
 
 // handles sharding OffsetFetchRequest
 type offsetFetchSharder struct{ *Client }
-
-func (*offsetFetchSharder) unpackPinReq() bool { return true } // batch first, single group fallback
 
 func offsetFetchReqToGroup(req *kmsg.OffsetFetchRequest) kmsg.OffsetFetchRequestGroup {
 	g := kmsg.NewOffsetFetchRequestGroup()
@@ -2458,8 +2448,6 @@ func (*offsetFetchSharder) merge(sresps []ResponseShard) (kmsg.Response, error) 
 // handles sharding FindCoordinatorRequest
 type findCoordinatorSharder struct{ *Client }
 
-func (*findCoordinatorSharder) unpackPinReq() bool { return true } // batch first, single key fallback
-
 func findCoordinatorRespCoordinatorIntoResp(c kmsg.FindCoordinatorResponseCoordinator, into *kmsg.FindCoordinatorResponse) {
 	into.NodeID = c.NodeID
 	into.Host = c.Host
@@ -2477,9 +2465,12 @@ func (*findCoordinatorSharder) shard(_ context.Context, kreq kmsg.Request, lastE
 	req = &dup
 
 	uniq := make(map[string]struct{}, len(req.CoordinatorKeys))
-	uniq[req.CoordinatorKey] = struct{}{}
-	for _, key := range req.CoordinatorKeys {
-		uniq[key] = struct{}{}
+	if len(req.CoordinatorKeys) == 0 {
+		uniq[req.CoordinatorKey] = struct{}{}
+	} else {
+		for _, key := range req.CoordinatorKeys {
+			uniq[key] = struct{}{}
+		}
 	}
 	req.CoordinatorKeys = req.CoordinatorKeys[:0]
 	for key := range uniq {
@@ -2546,8 +2537,6 @@ func (*findCoordinatorSharder) merge(sresps []ResponseShard) (kmsg.Response, err
 
 // handles sharding DescribeGroupsRequest
 type describeGroupsSharder struct{ *Client }
-
-func (*describeGroupsSharder) unpackPinReq() bool { return false }
 
 func (cl *describeGroupsSharder) shard(ctx context.Context, kreq kmsg.Request, _ error) ([]issueShard, bool, error) {
 	req := kreq.(*kmsg.DescribeGroupsRequest)
@@ -2636,8 +2625,6 @@ func (*describeGroupsSharder) merge(sresps []ResponseShard) (kmsg.Response, erro
 // handles sharding ListGroupsRequest
 type listGroupsSharder struct{ *Client }
 
-func (*listGroupsSharder) unpackPinReq() bool { return false }
-
 func (cl *listGroupsSharder) shard(ctx context.Context, kreq kmsg.Request, _ error) ([]issueShard, bool, error) {
 	req := kreq.(*kmsg.ListGroupsRequest)
 	return cl.allBrokersShardedReq(ctx, func() kmsg.Request {
@@ -2666,8 +2653,6 @@ func (*listGroupsSharder) merge(sresps []ResponseShard) (kmsg.Response, error) {
 
 // handle sharding DeleteRecordsRequest
 type deleteRecordsSharder struct{ *Client }
-
-func (*deleteRecordsSharder) unpackPinReq() bool { return false }
 
 func (cl *deleteRecordsSharder) shard(ctx context.Context, kreq kmsg.Request, _ error) ([]issueShard, bool, error) {
 	req := kreq.(*kmsg.DeleteRecordsRequest)
@@ -2786,8 +2771,6 @@ func (*deleteRecordsSharder) merge(sresps []ResponseShard) (kmsg.Response, error
 // handle sharding OffsetForLeaderEpochRequest
 type offsetForLeaderEpochSharder struct{ *Client }
 
-func (*offsetForLeaderEpochSharder) unpackPinReq() bool { return false }
-
 func (cl *offsetForLeaderEpochSharder) shard(ctx context.Context, kreq kmsg.Request, _ error) ([]issueShard, bool, error) {
 	req := kreq.(*kmsg.OffsetForLeaderEpochRequest)
 
@@ -2904,8 +2887,6 @@ func (*offsetForLeaderEpochSharder) merge(sresps []ResponseShard) (kmsg.Response
 
 // handle sharding WriteTxnMarkersRequest
 type writeTxnMarkersSharder struct{ *Client }
-
-func (*writeTxnMarkersSharder) unpackPinReq() bool { return false }
 
 func (cl *writeTxnMarkersSharder) shard(ctx context.Context, kreq kmsg.Request, _ error) ([]issueShard, bool, error) {
 	req := kreq.(*kmsg.WriteTxnMarkersRequest)
@@ -3093,8 +3074,6 @@ func (*writeTxnMarkersSharder) merge(sresps []ResponseShard) (kmsg.Response, err
 // handle sharding DescribeConfigsRequest
 type describeConfigsSharder struct{ *Client }
 
-func (*describeConfigsSharder) unpackPinReq() bool { return false }
-
 func (*describeConfigsSharder) shard(_ context.Context, kreq kmsg.Request, _ error) ([]issueShard, bool, error) {
 	req := kreq.(*kmsg.DescribeConfigsRequest)
 
@@ -3160,8 +3139,6 @@ func (*describeConfigsSharder) merge(sresps []ResponseShard) (kmsg.Response, err
 // handle sharding AlterConfigsRequest
 type alterConfigsSharder struct{ *Client }
 
-func (*alterConfigsSharder) unpackPinReq() bool { return false }
-
 func (*alterConfigsSharder) shard(_ context.Context, kreq kmsg.Request, _ error) ([]issueShard, bool, error) {
 	req := kreq.(*kmsg.AlterConfigsRequest)
 
@@ -3224,8 +3201,6 @@ func (*alterConfigsSharder) merge(sresps []ResponseShard) (kmsg.Response, error)
 
 // handles sharding AlterReplicaLogDirsRequest
 type alterReplicaLogDirsSharder struct{ *Client }
-
-func (*alterReplicaLogDirsSharder) unpackPinReq() bool { return false }
 
 func (cl *alterReplicaLogDirsSharder) shard(ctx context.Context, kreq kmsg.Request, _ error) ([]issueShard, bool, error) {
 	req := kreq.(*kmsg.AlterReplicaLogDirsRequest)
@@ -3372,8 +3347,6 @@ func (*alterReplicaLogDirsSharder) merge(sresps []ResponseShard) (kmsg.Response,
 // handles sharding DescribeLogDirsRequest
 type describeLogDirsSharder struct{ *Client }
 
-func (*describeLogDirsSharder) unpackPinReq() bool { return false }
-
 func (cl *describeLogDirsSharder) shard(ctx context.Context, kreq kmsg.Request, _ error) ([]issueShard, bool, error) {
 	req := kreq.(*kmsg.DescribeLogDirsRequest)
 
@@ -3489,8 +3462,6 @@ func (*describeLogDirsSharder) merge(sresps []ResponseShard) (kmsg.Response, err
 // handles sharding DeleteGroupsRequest
 type deleteGroupsSharder struct{ *Client }
 
-func (*deleteGroupsSharder) unpackPinReq() bool { return false }
-
 func (cl *deleteGroupsSharder) shard(ctx context.Context, kreq kmsg.Request, _ error) ([]issueShard, bool, error) {
 	req := kreq.(*kmsg.DeleteGroupsRequest)
 
@@ -3577,8 +3548,6 @@ func (*deleteGroupsSharder) merge(sresps []ResponseShard) (kmsg.Response, error)
 // handle sharding IncrementalAlterConfigsRequest
 type incrementalAlterConfigsSharder struct{ *Client }
 
-func (*incrementalAlterConfigsSharder) unpackPinReq() bool { return false }
-
 func (*incrementalAlterConfigsSharder) shard(_ context.Context, kreq kmsg.Request, _ error) ([]issueShard, bool, error) {
 	req := kreq.(*kmsg.IncrementalAlterConfigsRequest)
 
@@ -3641,8 +3610,6 @@ func (*incrementalAlterConfigsSharder) merge(sresps []ResponseShard) (kmsg.Respo
 
 // handle sharding DescribeProducersRequest
 type describeProducersSharder struct{ *Client }
-
-func (*describeProducersSharder) unpackPinReq() bool { return false }
 
 func (cl *describeProducersSharder) shard(ctx context.Context, kreq kmsg.Request, _ error) ([]issueShard, bool, error) {
 	req := kreq.(*kmsg.DescribeProducersRequest)
@@ -3752,8 +3719,6 @@ func (*describeProducersSharder) merge(sresps []ResponseShard) (kmsg.Response, e
 // handles sharding DescribeTransactionsRequest
 type describeTransactionsSharder struct{ *Client }
 
-func (*describeTransactionsSharder) unpackPinReq() bool { return false }
-
 func (cl *describeTransactionsSharder) shard(ctx context.Context, kreq kmsg.Request, _ error) ([]issueShard, bool, error) {
 	req := kreq.(*kmsg.DescribeTransactionsRequest)
 
@@ -3839,8 +3804,6 @@ func (*describeTransactionsSharder) merge(sresps []ResponseShard) (kmsg.Response
 
 // handles sharding ListTransactionsRequest
 type listTransactionsSharder struct{ *Client }
-
-func (*listTransactionsSharder) unpackPinReq() bool { return false }
 
 func (cl *listTransactionsSharder) shard(ctx context.Context, kreq kmsg.Request, _ error) ([]issueShard, bool, error) {
 	req := kreq.(*kmsg.ListTransactionsRequest)

--- a/pkg/kgo/client_test.go
+++ b/pkg/kgo/client_test.go
@@ -1,7 +1,10 @@
 package kgo
 
 import (
+	"context"
+	"strconv"
 	"testing"
+	"time"
 
 	"github.com/twmb/franz-go/pkg/kmsg"
 )
@@ -103,4 +106,20 @@ func TestParseBrokerAddrErrors(t *testing.T) {
 			}
 		})
 	}
+}
+
+func TestUnknownGroupOffsetFetchPinned(t *testing.T) {
+	req := kmsg.NewOffsetFetchRequest()
+	req.Group = "unknown-" + strconv.FormatInt(time.Now().UnixNano(), 10)
+
+	cl, _ := NewClient(
+		getSeedBrokers(),
+	)
+	defer cl.Close()
+	defer func() {
+		if err := recover(); err != nil {
+			t.Errorf("fetch panicked: %v", err)
+		}
+	}()
+	req.RequestWith(context.Background(), cl)
 }

